File: interpreted/src/token.rs
use std::fmt::{Debug, Display};

use crate::literal::Literal;

#[derive(Debug, Copy, Clone, PartialEq)]
pub enum TokenKind {
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Bang,
    Ne,
    Eq,
    EqEq,
    Gt,
    Ge,
    Lt,
    Le,
    Identifier,
    String,
    Number,
    And,
    Class,
    Else,
    True,
    False,
    Fn,
    For,
    If,
    None,
    Or,
    Print,
    Return,
    Super,
    This,
    Var,
    While,
    Eof,
}

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

#[derive(Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub literal: Literal,
    pub line: usize,
}

impl Token {
    pub fn new(kind: TokenKind, lexeme: String, literal: Literal, line: usize) -> Self {
        Self {
            kind,
            lexeme,
            literal,
            line,
        }
    }
}

impl Debug for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "\"{}\" {}::({:?})", self.lexeme, self.kind, self.literal)
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.lexeme)
    }
}

File: js/src/token.ts
import type { TokenKind } from './token_kind';

export type Literal = boolean | number | string | null;

export class Token {
  public readonly kind: TokenKind;
  public readonly lexeme: string;
  public readonly literal: Literal;
  public readonly line: number;

  constructor(kind: TokenKind, lexeme: string, literal: Literal, line: number) {
    this.kind = kind;
    this.lexeme = lexeme;
    this.literal = literal;
    this.line = line;
  }

  // eslint-disable-next-line @typescript-eslint/naming-convention
  public toString(): string {
    return this.kind + ' ' + this.lexeme + ' ' + this.literal;
  }
}
